coremltools/converters/mil/frontend/_utils.py

```diff
@@ -223,6 +223,38 @@ def get_output_names(outputs):
     return output_names
 
 
+def solve_diagonal_einsum(parsed_vectors, vars):
```

Comment:
Do you think it would be a better implementation if we first group those diagonal dimensions together at once?
For instance,
a with shape `(2, 3, 3, 2, 3, 4)`, with the first operation `i j j i j k`,
we have a group of `i` and `j`

So we can do a single `tranpose` op with `perm = [0,3,1,2,4,5]`,
and two `gather_nd` ops with indices
`[0,0],[1,1]` and `[0,0,0],[1,1,1],[2,2,2]`,
this would streamline the ops decomposition imo.

Reply:
I changed to handle equation `"aaa"` by one `gather_nd`.
In this sense, I reduced `transpose` and `gather_nd`.

But in case with "abab", I have no idea to reduce `transpose` and `gather_nd`.

`"abab"` -> 1st transpose -> `"aabb"` -> 1st gather -> `"abb"` -> 2nd transpose -> `"bba"` -> 2nd gather `"ba"`

In my thought, since `gather_nd` indices specifies first axes, duplicated axes should be place at first.
So 2nd transpose is needed for this reason.
To handle `"abb"` without additional `transpose`, we need `batch_dims` option for `gather_nd`.

https://www.tensorflow.org/api_docs/python/tf/gather_nd